### ISSUE TITLE
nth() (and therefore head()) to support Iterators

### ISSUE THREAD

**Issue by ukslim** [ramda/ramda#2508]
I'm aware that support for ImmutableJS has been discussed and deemed out of scope: https://github.com/ramda/ramda/issues/1367

However I've noticed that a lot of Ramda works with Immutables' Seq (etc) because Ramda detects that the List is an Iterable. `map()` and `find`, for example, work fine:

    const double = compose(
         x => x * 2,
         tap(() => console.log('called')
    );

    compose(
        find(equals(8)),
        map(double)
    )(immutables.Range());

... logs 5 times then returns `8`.

However, while this returns 0:

    compose(head, map(double))([0,1,2])

This returns `undefined`:

   compose(head, map(double))(immutables.Range(0,3))

... because `head` calls `nth(0)` and `nth` only caters for arrays and strings.

Any comments or objections to my writing a PR to make `nth` (or possibly just `head`) understand iterables? 

I proved the concept with my own `head`:

```
const symIterator = typeof Symbol !== 'undefined' ? Symbol.iterator : '@@iterator';
const { head: oldHead } = require('ramda');

const head = list => {
  if (list[symIterator] != null) {
    return list[symIterator]().next().value;
  };
  return oldHead(list);
};
```

Of course there are more efficient ways to get the nth member of many objects that happen to be Iterables. In the case of an immutables Seq, `get(n)` -- but `iterable` is standardised and often we use low numbers in `nth`

**Comment by semmel** [comment]
I think it would be awkward if `R.head` works on iterables but not the other three `tail`, `init`, `last` of the quartet as well. 
Admitted the indices are low integers, but without a `ReverseIterator`in JS, it feels wrong even attempting it. Also with `R.nth` with negative index one would need to run to the end of the sequence in the implementation. 

For infinite sequences we'd get infinite loops, but this I'd not consider a big problem, since that's Ramda's principle "Garbage in - garbage out".


While I like JS iterables and think that many of Ramda's List functions are in principle suitable candidates (`contains`, `startsWith`, ...)  to work on iterable sequences I think JS iterables do not qualify as full List collection type - it's just an interface for **looping/iterating** of the sequence.

**Comment by ukslim** [comment]
As it happens, `tail` works on an *immutables* object, because Ramda dispatches to the object's `slice`. Hence with my `head` implementation `head(R.tail(imm.Range(5,Infinity)))` returns 6 as you'd want. Ramda is so close to working with these kinds of list.

Is there a preferable interface to Iterable that `head` could use? It could itself use `slice()` of course.

**Comment by ukslim** [comment]
(I'm not especially concerned about `init`/`last` because I have a left-over instinct from learning Lisp all those years ago, thinking of the list as a linked-list. `head` and `tail` (`car`/`cdr`) are super-efficient; `last` involves traversing the whole list.

**Comment by semmel** [comment]
It may be I am the wrong guy to answer, since I am not familiar ImmutableJS. I've assumed you target [JavaScript Iterables](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols). That's why I am confused:
> Is there a preferable interface to Iterable that head could use?

The only interfaces there are is `[Symbol.iterator]` for Iterables and `next()` for Iterators. Nothing else.

I don't know what `imm.Range(5, Infinity)` is, but this is an infinite JS iterable
```javascript
function * naturalNumbersFrom(n){
    yield n;
    yield * naturalNumbersFrom(n + 1);
}

var it = naturalNumbersFrom(10);
it.next().value; // -> 10
it.next().value; // -> 11
...
```
for which `R.tail` would be a futile attempt indeed.

**Comment by ukslim** [comment]
@semmel what I mean about interfaces might be clouded by nomenclature.

An object can have a bunch of functions as members. `R.tail` happens to work on an immutable.js `Seq` -- even if it's a `Range` with no end, because `R.tail` calls `R.slice(1,Infinity)`, and `R.slice` dispatches to its parameter's `slice` method if there is one.

So `slice` and `tail` operate on any object that conforms to what I'll call `Sliceable`.

Many of Ramda's functions dispatch to a load of member functions -- `slice`, `reduce`, `map`, `chain` -- the list goes on.

But `head` does not. It works on a String or an array, and that's it.

Now, I know the immutable.js types implement Iterable, so I can make it work by implementing `head` in terms of the iterator. In the case of immutable.js I could also make it work by dispatching to the object's `first` method. That's an interface I might call `Firstable`, which Ramda currently doesn't recognise.

So I'm asking if there's an interface (maybe better to use the word 'specification' as FantasyLand does) other than Iterable, which Ramda's `head` could reasonably dispatch to.

**Comment by CrossEye** [comment]
If Ramda expanded its dispatching, it's possible that we could offer full integration with [funkia/list][fl].  But that's an extremely controversial idea.  Some of the core team would rather remove as much of it as we can.

Although I understand the objection from @semmel, I think that we could reasonably make `head` and `tail` work on iterables even if `last` and `init` don't.  Or we could create them and simply let them fail if they're passed infinite sequences.  People who try to get `last` on an infinite list deserve to wait a while! :smile:


  [fl]: https://github.com/funkia/list#seamless-ramda-integration

**Comment by ukslim** [comment]
Here's a branch for discussion: https://github.com/ukslim/ramda/tree/nth-for-iterable

**Comment by Bradcomp** [comment]
I have a concern with regards to immutability and iterators.

Iterators are inherently stateful. An iterator gets consumed as one calls `next` on it. Thisd leads to some rather unintuitive behavior when working with these functions:
```
 const MyIterable = function*(){ var n = 0; while(true) yield n++; }
undefined
> const myIterator = MyIterable()
undefined
> myIterator.next()
{ value: 0, done: false }
> const head = (it) => it.next().value;
undefined
> const tail = (it) => { it.next(); return it; };
undefined
> const [a, b] = [head(myIterator), head(tail(myIterator))];
undefined
> a
1
> b
3

head(myIterator) === head(myIterator) // false

> const [c, d] = [head(tail(myIterator)), head(myIterator)]
undefined                                                  
> c                                                        
7                                                          
> d                                                        
8                                                          
```

and so on. This is a ll expected behavior, but it _doesn't_ mesh well with the principles of Ramda as a whole.

**Comment by CrossEye** [comment]
@Bradcomp: Right.  It seems we'd have to implement something like [`immutagen`][im]  in order to get these to work as we'd like, and that seems a great deal of overhead.


  [im]: https://github.com/pelotom/immutagen

**Comment by semmel** [comment]
```javascript
var init = function(list){
    var itbl = {};
    itbl[Symbol.iterator] = function(){
        const it = list[Symbol.iterator]();
        var currentResult = it.next();
        return { next: function(){
		var result = currentResult; 
		const nextResult = currentResult.done ? currentResult: it.next();
		if (nextResult.done){
			result = {value: undefined, done: true};
                }
		currentResult = nextResult;
		return result;
		}
 	};
    };
    return itbl;
};

var head = list => list[Symbol.iterator]().next().value;

var ar = [0, 1, 2, 3];
for (const x of init(ar)){ console.log(x); } // -> 0, 1, 2

head(init(init(init(ar)))); // -> 0
head(init(init(init(init(ar))))) // -> undefined

head(init(init(myIterable))); // -> 0
```
`last` should be trivial.

@ukslim could you perhaps run my ad.hoc implementations against ImmutableJS?

So it would be useful to extend some of Ramda's List functions on iterables? 
Should attempt a PR?

**Comment by CrossEye** [comment]
@semmel: That makes sense.  I think a PR might be a fair bit of work.  While I'm in favor of seeing this, I would love for more of @ramda/core to chime in before you put in that effort.  (Of course, if you're itching to get started...)

**Comment by ukslim** [comment]
@semmel Yes, your implementation of `tail` works:

```
var { Seq } = require('immutable');
head(init(init(Seq([0,1,2,3])))); // -> 0
head(init(init(init(Seq([0,1,2,3]))))); // -> 0
head(init(init(init(init(Seq([0,1,2,3])))))); // -> undefined
```

And yes, you're correct that statefulness isn't a concern for an `Iterable` -- the implementation requests a new iterator (which is stateful), uses it to iterate, then discards it without ever exposing it to the caller.

**Comment by paldepind** [comment]
@ukslim 

> funkia/list is plausibly usable, however, it doesn't really work with (all of) Ramda's functions. Rather it provides its own functions with the same names as Ramda's, operating on its own datatype, hence the example on its front page:

That is true (but List does implements Foldable, Functor and Iterable which are abstractions that Ramda supports). The big benefit of the approach that List takes is that it actually works _reliably_ with Ramda.

We can make `head` work for any iterable and we could also implement a very slow `nth` for any iterable. But very soon one runs into limitations. For instance, you can't implement `insert`, `filter`, `partition`, `zip`, `sort`, and many many more methods with Iterable. This means that attempts to make Ramda work directly on other data-structures only works every now and then. End users will, for each case, have to remember if a particular function works or not. That is not to say that doing that is a bad idea. It definitely has value. But the approach List takes has other benefits IMO. It can actually work 1-to-1 in a predictable manner and without sacrificing performance.

Ramda already supports iterables in a lot of functions which is nice. Making `head` work on iterables seems reasonable as well. But, I think making `nth` work on iterables would have bad consequences. Such an implementation would have a big performance overhead and people may not realize that they're writing really slow code. For instance, users might be tempted to call `R.nth` on a list from List without knowing that it is a lot slower than the `nth` function that List exports. That would slow down users code with no benefits.

In other words. I think supporting Iterables is great when it doesn't have a huge performance overhead. When it does have a huge performance overhead I'm afraid it will do more harm than good.

**Comment by semmel** [comment]
First let me focus the discussion on [JS Iterables](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols) and put ImmutableJS and funkia/list aside.

@paldepind states that there are at least two drawbacks, which I agree with:
1.) A documentation issue.
> ... attempts to make Ramda work directly on other data-structures only works every now and then. End users will, for each case, have to remember if a particular function works or not.

If I understand JS Iterables as forward linked lists I cannot say which of Ramda's List functions can't (or shouldn't) be implemented. I don't agree that e.g. `filter` can't. (See example below)

2.) Performance. 
as @ukslim said
> ... thinking of the list as a linked-list. head and tail (car/cdr) are super-efficient; last involves traversing the whole list

Hmm. One could argue that if you put your data into a linked list the user should be aware of the performance implications. So it's his/her own fault according to Ramda's famous principle ("Garbage in - garbage out").

```javascript
// just proof-of-concept
var filter = function(pred, list){
    var tailItbl = {};
    tailItbl[Symbol.iterator] = function(){
        const it = list[Symbol.iterator]();
		
        return { next: function(){
		var itResult = it.next(); 
		while (!itResult.done && !pred(itResult.value)){
			itResult = it.next();
            	};
		return itResult;
	}};
    };
    return tailItbl;
};

var head = list => list[Symbol.iterator]().next().value;
var myIterable = {};
myIterable[Symbol.iterator] = function*(){ var n = 0; while(true) yield n++; };

head(filter(R.gt(R.__, 1), myIterable)); // -> 2
head(filter(x => x % 5 === 0, filter(R.gt(R.__, 11), myIterable))); // -> 15
```

**Comment by paldepind** [comment]
@semmel 

> First let me focus the discussion on JS Iterables and put ImmutableJS and funkia/list aside.

I think that is a good idea :smile: But, we still need to keep in mind that Iterables are not _just_ iterables. They're an abstraction implemented by data-structures that are a lot of other things in addition to being iterables. Iterables are like functors and foldables—they're an abstraction.

> If I understand JS Iterables as forward linked lists I cannot say which of Ramda's List functions can't (or shouldn't) be implemented.

There is a connection between Iterables and forward linked list in that the performance of doing something on a linked list is similar to the performance of doing the same thing on an iterable. `head` is O(1), `nth` is O(n), etc. But, JS Iterables are not forward linked lists. They are data-structures that support iteration through a specified API. A lot of data-structures can implement the Iterable abstraction without being linked lists. For instance, JS Arrays are Iterables and List's relaxed radix balanced trees are also Iterables.

> One could argue that if you put your data into a linked list the user should be aware of the performance implications.

People put their data into a linked list. People put their data into data-structures that, among other things, happens to be an Iterable.

> I don't agree that e.g. filter can't. (See example below)

While the `filter` you've defined works well it has the problem that it returns a different type than the one given to `filter`. `filter` normally returns an instance of the same data-structure that it was given. I.e. filtering and array results in another array. But the `filter` you've defined always returns a data type which is nothing but an Iterable.

**Comment by semmel** [comment]
@paldepind 
Let me explain. The idea was *not to replace the existing* implementations of the List functions, but to *add* support for iterables as fallback. So the implementation would proceed from *dispatching* (e.g. to a `filter`method if the data should happen to provide it), handling expected List types (`isArray`) - all like it is now - to finally inspect the provided List `Object` for iterator support.

>For instance, JS Arrays are Iterables and List's relaxed radix balanced trees are also Iterables.

Yes if it happens to be a JS array, we can employ array methods, if it happens to be a List tree, it's api should be employed. But if it is neither of these types we could try the very general abstraction of a JS iterable.

However my plan has a problem for `filter`, `reject` and so one which work on `Filterables` which can be plain objects too. So if that key-value collection `Filterable` also is an interable; Should `filter` work on the key-value pairs at it's right now or employ the iterator from the collection?

> filter normally returns an instance of the same data-structure that it was given. ... But the filter you've defined always returns a data type which is nothing but an Iterable.

Yes, the same is true for the `tail`and `init` examples above - they are not meant to be comprehensive. I left this part out. Of course the given data structure must be cloned *(shallow?)* in the implementation and returned as result. The only change would be the clone's iterator which swallows the filter predicate.

I don't know if this all is perfectly reasonable, therefore I am grateful for the good discussion. 😄

**Comment by jmagaram** [comment]
A general programming principle is that a function should not over-specify input parameter requirements because that restricts the usefulness of the function. Input arguments should be the most general and abstract possible. For example a “get length” function should not require a Set because that prevents usage with Array, Map, String, etc.

Output results should be the most specific since they are the most useful. For example if the user asks for Unique items then that should be returned as a Set, not just an Iterable, because sets already have an index to quickly find values while Iterables and Arrays do not.

Most general abstract in. Most specifically useful out.

Many Ramda functions over-specify the input to an Array when any kind of Iterable will do. Requiring an Array for input decreases runtime performance because the caller needs to first materialize a whole array when one-at-a-time processing is sufficient. Ramda would be more useful if it accepted Iterables everywhere it could. Nth, Head, Aperture, None, Length, Median, and others. If the input object is an Array or other well-known object with indexed access then certainly go ahead and use that optimization. But as a fallback I see no reason, other than download size, why all these functions shouldn’t work with Iterables or have iterable variants. Iterables are immutable (good) and lazy (good). 

Likewise Ramda over specifies output results to Array for no obvious benefit and with the performance cost of materializing one. Much better to stream list results as an Iterable until the user says “ok I’m done with my filtering, mapping, inserting, expanding, and flattening- now give me an Array.” Backward compatibility aside, It would be nice to see more Ramda functions output Iterables, since lazy computations are efficient.

**Comment by CrossEye** [comment]
@jmagaram:

I think we're all pretty well in agreement that this is worth doing.  I've wanted this ability for years (thinking in terms of more general streams than specifically Iterables, but it's the same principle.)  We've discussed some possibilities since at least #446.

But as mentioned over on #2525, there is quite a bit of work to do here, both in API design and in implementation.  Some API questions:

  - Given this as your data:

    ```js
    const foo = {
      [Symbol.iterator]: function* f(n = 0) {
        yield n
        yield * f(n + 1)
      }
    }
    ```

    What is the expected behavior of some of those functions you mention, e.g.  `none`, `length`, `median`? (`nth`, `head`, and `aperture` seem to have straightforward answers.)  Ramda's usual philosophy of "let it fail" is much more problematic when failing means that your application crashes... and you can't even catch an exception to handle it.

  - Would we expect to be able to get back a more specific type than just Iterable, if we supplied data of a more specific type?  What is the return type of

     ```js
     map(n => n * n, new MyStream(/* ... */))
    ```

     if instances of MyStream have `Symbol.iterator`?  `Iterable`?  `MyStream`? `map`'s usual contract is `map :: Functor f => (a -> b) -> f a -> f b`.  So the output type is an instance of the same Functor as the input type.  Would we have to forgo this?  Do we try to apply some gloss that says, "Well, `Iterable` is our Functor here, not `MyStream`?"  Do we depend upon delegation to handle this before our iterable branch?  Something else?

  - How do we choose the functions that need this treatment?  Iterables are not arrays, but they do overlap a great deal.  Still, there are functions which make sense on arrays that might not on iterables.  Is there some clear criteria we can use to choose when to add a code-path for iterables to existing functions?


> Likewise Ramda over specifies output results to Array for no obvious benefit and with the performance cost of materializing one.

Most of Ramda's transformation functions will always return the same type as is input to them.  This is important IMHO to allow clear reasoning about a system.  I wouldn't want, for instance:

```js
const nbrs = [1, 2, 3, 4, 5]
const squares = map(n => n * n, nbrs) //=> iterarable yielding 1, 4, 9, 16, 25
squares[1] //=> undefined
```

That `[1]` is part of the Array interface, and Ramda should not be trying to enforce that you will need to use `nth(1)` on the results of a `map` call over an array.

But this is not any major issue; we can simply ensure that the array checks are made before the iterable ones.

The fact that Clojure can transform its collection types in similar circumstances is one of my main objections to it.  So I would suggest that if an array is input, Ramda should always output an array.

-----

By the way, wanting to do this is not an application of [Postel's Principle][pp] ("Be conservative in what you send, be liberal in what you accept.")  It is instead the recognition that there is a common abstraction underlying arrays, streams, iterables, etc., and that many of our functions could operate at that level.  Underscore, lodash, and particularly jQuery are built on Postel's Principle, but Ramda's underpinnings are a bit different.


  [pp]: https://en.wikipedia.org/wiki/Robustness_principle

**Comment by jmagaram** [comment]
If you give an infinite sequence to a function that requires a finite sequence, like length or sort, then the software should fail. This is a programmer error, like passing a negative number to square root. Infinite sequences are useful, like a random number generator, as long as you know when to stop pulling values.

If you have a function that takes an X as input I don't think the function can always generate an X on output. Otherwise the function would need to know too much - how to build a Map, a Set, an immutable.js Set, Stack, a custom collection, etc. and this is clearly not feasible. It also may not be necessary if the user wants to take the result and do further stream processing on it before materializing it. Better to output an Iterable and let the caller construct an X once they are ready for it. This lazy behavior will result in better performance. Of course if the function generates an Array or Set or Map to do its job, then return that rather than throwing away the extra info and just returning an Iterable.

I am very comfortable with code in C# that takes an array, does a bunch of stream processing on it with iterables, and then converts it back to an array when I'm done. That actually feels really good to me. My first reaction when I read the Ramda docs was "wow this is all really cool and powerful, but why create all these intermediate arrays?"

I think you could add Iterable input to nearly every function that takes an array. I don't know what to do about output. If you were starting from scratch you could output iterables everywhere. So maybe for backward compatibility it makes sense to have alterate versions that output iterables rather than arrays and the user of the library can pick which one they want.

**Comment by paldepind** [comment]
@CrossEye 

> Most of Ramda's transformation functions will always return the same type as is input to them. This is important IMHO to allow clear reasoning about a system.

I very much agree.

@jmagaram 

> If you have a function that takes an X as input I don't think the function can always generate an X on output. Otherwise, the function would need to know too much - how to build a Map, a Set, an immutable.js Set, Stack, a custom collection, etc. and this is clearly not feasible.

I think it just requires the right abstraction. For instance `map` can always return the same output thanks to the Functor abstraction. Iterables, on the other hand, is not a good abstraction for `map` as there are many things that can support `map` but which cannot implement Iterable (async streams for instance).

IMO Iterable is a fine abstraction for functions that are essentially just iterations over a data structure (`R.find`, `R.mean`, etc). But, it is a poor abstraction for many other things like `R.map`, `R.nth`, `R.filter`, etc).

Sure, we _could_ accept any Iterable in `R.last` and `R.nth` but I don't think that is necessarily a good idea. To me, it feels like putting a screw into a piece of wood with a hammer in the conviction that a hammer is a very general purpose tool.

The very first thing we should do in Ramda is to properly document when our functions supports Iterables. There are many functions that currently support Iterables with no mention in the documentation. Fixing that would be an obvious improvement to the current situration.

**Comment by CrossEye** [comment]
@paldepind:

While I agree overall, there are several things that I'm not so sure of:

> Iterables, on the other hand, is not a good abstraction for `map` as there are many things that can support `map` but which cannot implement Iterable (async streams for instance).

I absolutely agree that the data supplied to `map` cannot always implement Iterable, would you see any problem with `map` supporting Iterable as a last check, as I mention in my previous comment?

> But, [Iterable] is a poor abstraction for many other things like R.map, R.nth, R.filter, etc.

What do you mean by "is a poor abstraction for" here?  If you're suggesting that `map`, `filter`, *et al* have more general uses than simply on iterables, then I heartily agree.  But if you're suggesting that there are not reasonable functions to apply to iterables, then I don't agree.  Of course `nth` will have vastly different performance characteristics than it does on an array, and it's possible to imagine iterable data structures on which it has no real meaning, depending on exactly how iterable is defined.  But the others seem to have reasonable meanings, and straightforward implementations.

> There are many functions that currently support Iterables with no mention in the documentation. Fixing that would be an obvious improvement to the current situration.

Absolutely.

**Comment by jmagaram** [comment]
I think I'm starting to understand more how Ramda works. I just read up on "Functors" and FantasyLand. It seems like Ramda works with abstract data types defined by FantasyLand, not just specific ones like Array. So if you use the filter function on something of type X you want to get an X back, just like if you do number+number you want to get a number back, and string+string you want to get a string back. Correct? Ok that makes sense to me and is quite elegant.

So maybe I could wrap any iterable in a lightweight wrapper, give it a bunch of fantasyland methods to make it a Functor (map), Filterable (filter), Setoid (equals), Semigroup (concatenation), Monoid (empty sequence), Traverseable (?), and then use Ramda on it. Because as I look through the FantasyLand specs it seems like most of those capabilities apply to any kind of iterable. Maybe this "wrapper" should be part of Ramda.

But as I look through the Ramda docs it doesn't look like Ramda consistently works with abstract data types. Head and Nth, for example, take an array or string. Why not take anything that is fantasyland "traversable" and return the result from doing that 1 or N times? Maybe there is a way to get Zip, UniqBy, Aperature, etc. to work with abstract data types like iterable too.

**Comment by CrossEye** [comment]
@jmagaram:

> It seems like Ramda works with abstract data types defined by FantasyLand, not just specific ones like Array.

To some extent.  Ramda started working with arrays, objects, strings, and functions well before we tried integrating with FantasyLand.  This integration is quite important to Ramda, but it is not the only thing that Ramda is about.

> So maybe I could wrap any iterable in a lightweight wrapper, give it a bunch of fantasyland methods...

Sure, but you could, but you might lose some specific features of your underlying type in doing this.  That's why we're concerned about maintaining the input types.  Since this works:

```js
[1, 2, 3].map(n => n * n).push(42) //=> [1, 4, 9, 42]
```

we would also expect this to work:

```js
R.map(n => n * n, [1, 2, 3]).push(42) //=> [1, 4, 9, 42]
```

which wouldn't happen if `R.map` returned a lightweight container.


> [I]t doesn't look like Ramda consistently works with abstract data types.

Absolutely not.  We have plenty of functions, and only a small number of them work on multiple types.  But those tend to be the most important in the library.


> Maybe there is a way to get Zip, UniqBy, Aperature, etc. to work with abstract data types like iterable too.

**Quick note::: I know you're coming from .NET, but function names in Javascript, and all of Ramda's functions, are written in `lowerCamelCase`.

You should note that the FantasyLand types are themselves abstract, and generally more abstract than is Iterable.  (You can easily make any Iterable a Functor, for instance, but there are plenty of Functors which are not iterable.)  `zip` and `aperture` could easily work with iterables.  I'm less convinced about `uniq` or `uniqBy`; there might be some difficulties again with infinite iterables.

----------

By the way, I think you may have a misapprehension of `traverse`.  Tom Harding's [explanation][tr] is worth a read although you may have to read earlier parts of his excellent [series][th] to follow all of it.


  [tr]: http://www.tomharding.me/2017/05/08/fantas-eel-and-specification-12/
  [th]: http://www.tomharding.me/fantasy-land/

**Comment by jmagaram** [comment]
I just tried reading about Traversable and my head exploded. This is too abstract for me. Maybe FantasyLand doesn't have the concept of a "collection" that can be traversed. Iterable is a really nice abstract data type. There's got to be a way to get it to play nice with Ramda. A wrapped Iterable can definitely be a fantasyland Foldable. head, nth, maxBy, etc. can be calculated using a fold/reduce method. Aperature is some kind of map - map a collection<T> to a collection<T[]>. I think I'm going to bow out here.
